commit e4e794841e3140875f2aa86b90e2ada3d61e1244
Author: Florian Weimer <fweimer@redhat.com>
Date:   Fri Apr 7 13:01:21 2017 +0200

    resolv: Remove IQUERY support

diff --git a/resolv/res_mkquery.c b/resolv/res_mkquery.c
index 5a0bb1044b55a643..8279d15de4271c84 100644
--- a/resolv/res_mkquery.c
+++ b/resolv/res_mkquery.c
@@ -188,24 +188,6 @@ res_nmkquery(res_state statp,
 		hp->arcount = htons(1);
 		break;
 
-	case IQUERY:
-		/*
-		 * Initialize answer section
-		 */
-		if (__glibc_unlikely (buflen < 1 + RRFIXEDSZ + datalen))
-			return (-1);
-		*cp++ = '\0';	/* no domain name */
-		NS_PUT16 (type, cp);
-		NS_PUT16 (class, cp);
-		NS_PUT32 (0, cp);
-		NS_PUT16 (datalen, cp);
-		if (datalen) {
-			memcpy(cp, data, datalen);
-			cp += datalen;
-		}
-		hp->ancount = htons(1);
-		break;
-
 	default:
 		return (-1);
 	}
