commit 363911ce1313a246b7d33f0983a09e7ab2525b3a
Author: Florian Weimer <fweimer@redhat.com>
Date:   Sat Jun 3 08:37:13 2017 +0200

    getaddrinfo: Eliminate another strdup call

diff --git a/sysdeps/posix/getaddrinfo.c b/sysdeps/posix/getaddrinfo.c
index 0e92221853dd73b1..317413c9ef1559e6 100644
--- a/sysdeps/posix/getaddrinfo.c
+++ b/sysdeps/posix/getaddrinfo.c
@@ -732,7 +732,7 @@ gaih_inet (const char *name, const struct gaih_service *service,
 			(*pat)->name = NULL;
 		      else if (canonbuf == NULL)
 			{
-			  canonbuf = strdup (air->canon);
+			  canonbuf = __strdup (air->canon);
 			  if (canonbuf == NULL)
 			    {
 			      result = -EAI_MEMORY;
