Emergency patch for broken utimes/utime functions.

diff --git a/sysdeps/unix/sysv/linux/utime.c b/sysdeps/unix/sysv/linux/utime.c
index 2cd9334a6f1efc36..6516344adcc16389 100644
--- a/sysdeps/unix/sysv/linux/utime.c
+++ b/sysdeps/unix/sysv/linux/utime.c
@@ -32,7 +32,7 @@ __utime64 (const char *file, const struct __utimbuf64 *times)
       ts64[1].tv_nsec = 0LL;
     }
 
-  return __utimensat64_helper (0, file, times ? ts64 : NULL, 0);
+  return __utimensat64_helper (AT_FDCWD, file, times ? ts64 : NULL, 0);
 }
 
 #if __TIMESIZE != 64
diff --git a/sysdeps/unix/sysv/linux/utimes.c b/sysdeps/unix/sysv/linux/utimes.c
index 75927b6ec6695c96..02a5e91415f52f2e 100644
--- a/sysdeps/unix/sysv/linux/utimes.c
+++ b/sysdeps/unix/sysv/linux/utimes.c
@@ -29,7 +29,7 @@ __utimes64 (const char *file, const struct __timeval64 tvp[2])
       ts64[1] = timeval64_to_timespec64 (tvp[1]);
     }
 
-  return __utimensat64_helper (0, file, tvp ? ts64 : NULL, 0);
+  return __utimensat64_helper (AT_FDCWD, file, tvp ? ts64 : NULL, 0);
 }
 
 #if __TIMESIZE != 64
