diff -Nrup a/resolv/nss_dns/dns-host.c b/resolv/nss_dns/dns-host.c
--- a/resolv/nss_dns/dns-host.c	2012-06-07 09:10:06.145448414 -0400
+++ b/resolv/nss_dns/dns-host.c	2012-06-07 09:18:40.850019678 -0400
@@ -745,6 +745,10 @@ getanswer_r (const querybuf *answer, int
 
       if ((qtype == T_A || qtype == T_AAAA) && type == T_CNAME)
 	{
+          /* A CNAME could also have a TTL entry.  */
+          if (ttlp != NULL && ttl < *ttlp)
+             *ttlp = ttl;
+
 	  if (ap >= &host_data->aliases[MAX_NR_ALIASES - 1])
 	    continue;
 	  n = dn_expand (answer->buf, end_of_message, cp, tbuf, sizeof tbuf);
@@ -906,7 +910,7 @@ getanswer_r (const querybuf *answer, int
 	    {
 	      register int nn;
 
-	      if (ttlp != NULL)
+              if (ttlp != NULL && ttl != 0 && ttl < *ttlp)
 		*ttlp = ttl;
 	      if (canonp != NULL)
 		*canonp = bp;
